### PR TITLE
Supporting log probabilities of prompt tokens in both engine and OpenAI API server (aka `echo`)

### PR DESCRIPTION
**Pull request by wanmok**
According to issue #201 , the `echo` function is not currently supported in vLLM. This implementation does not require making modifications to individual model files but lay the heavy work on the first generation step.

The basic logic of the implementation:
- Engine
  1. Added `echo` to the `SamplingParams` so that the engine can be informed of the `echo` request
  2. Modified `Sampler` to process the `echo` request.
    a. Flags from `SamplingParams` of each `SequenceGroup` would be used to determine whether to perform `echo`. Log probabilities of prompt tokens are computed at the first generation step. 
    b. Modified `SequenceOutputs` to carry prompt log probs from the first generate step so that engine can post process   `SequenceData` to store the generated log probabilities of prompt tokens.  Since log probabilities of prompt tokens are not necessarily stored in the top-k returns, they are stored in two separate fields as `prompt_logprobs` and `prompt_top_logprobs`.
    c. Changed numerical unstable computation of `log(softmax())` to `log_softmax()`.
  3. Modified `RequestOutput.from_seq_group` so that log probabilities of prompt tokens can be inserted into the final outputs.
- OpenAI API server
  1. Removed error message for requesting `echo`.
  2. Modified `create_logprobs` function to reflect engine output changes.
  3. Edge case: added the support for the case of `echo=True` and `max_tokens=0`, which is a valid case in OpenAI APIs. In this case, the `SamplingParams.max_tokens` would be set to `1` and enable a `echo_self` flag. The flag is then used in the following post processing to ensure the additionally generated token being removed from the final output.

Please let me know if you have any questions or concerns with respect to merging this PR. Thanks!

### CODE REVIEW

**Comment by eugenepentland** [comment]
Hello, thanks for this! Have you tested it using the openai compatible endpoint? when I try using curl it gets the request but I never get a response back. The following is with echo off, but it hangs regardless if it is on/off.

`
INFO 09-07 06:06:49 async_llm_engine.py:226] Received request cmpl-a06d218f31d549d7b128ee067691a83a: prompt: 'San Francisco is a', sampling params: SamplingParams(n=1, best_of=1, presence_penalty=0.0, frequency_penalty=0.0, temperature=0.0, top_p=1.0, top_k=-1, use_beam_search=False, length_penalty=1.0, early_stopping=False, stop=[], ignore_eos=False, max_tokens=7, logprobs=None), use_echo=False, prompt token ids: [1, 3087, 8970, 338, 263].
`

I may just be impatient, but I am excited to try this out with llm eval harness!

**Comment by wanmok** [comment]
> Hello, thanks for this! Have you tested it using the openai compatible endpoint? when I try using curl it gets the request but I never get a response back. The following is with echo off, but it hangs regardless if it is on/off.
> 
> `INFO 09-07 06:06:49 async_llm_engine.py:226] Received request cmpl-a06d218f31d549d7b128ee067691a83a: prompt: 'San Francisco is a', sampling params: SamplingParams(n=1, best_of=1, presence_penalty=0.0, frequency_penalty=0.0, temperature=0.0, top_p=1.0, top_k=-1, use_beam_search=False, length_penalty=1.0, early_stopping=False, stop=[], ignore_eos=False, max_tokens=7, logprobs=None), use_echo=False, prompt token ids: [1, 3087, 8970, 338, 263].`
> 
> I may just be impatient, but I am excited to try this out with llm eval harness!

What is the LM you were using for the request? Also, how many GPUs?

**Comment by eugenepentland** [comment]
Turns out I still had tensor_parallel=2 on, not sure the issue is related to your fork. Setting it to 1 fixed that issue. Running Llama-7B-HF on a RTX 3090. 

This is the new error internal server error I'm getting when I have the following args to http://localhost:8000/v1/completions with the openai endpoint:

JSON Args:
```
{
    "model": "llama",
    "prompt": "San Fransisco is a",
    "max_tokens": 0,
    "echo": true,
    "logprobs": 5
}
```

Error:
```
  File "/home/epentland/miniconda3/envs/axol/lib/python3.9/site-packages/fastapi/routing.py", line 190, in run_endpoint_function
    return await dependant.call(**values)
  File "/home/epentland/ai/inference/echo/vllm/vllm/entrypoints/openai/api_server.py", line 565, in create_completion
    logprobs = create_logprobs(
  File "/home/epentland/ai/inference/echo/vllm/vllm/entrypoints/openai/api_server.py", line 182, in create_logprobs
    for i, p in top_logprobs[i].items()
    ```

**Comment by eugenepentland** [comment]
I tested on the latest commit on my own hardware, and on a A6000 on runpod, getting the same error on both. Can you send the exact command for the openai ai sever and the curl request/python code you used to make the request?

**Comment by wanmok** [comment]
> I tested on the latest commit on my own hardware, and on a A6000 on runpod, getting the same error on both. Can you send the exact command for the openai ai sever and the curl request/python code you used to make the request?

Ah, I have it fixed in the latest commit. It was later introduced the modifications to the `top_logprobs`, where the first one should be `None`. Could you try to run it again with your dataset? This PR introduced multiple changes, so I suppose that it would be better to test more rigorously... Thanks for catching that.

**Comment by eugenepentland** [comment]
It's working on my end now! Thank you for the great work!

 I am making it work with lm-eval-harness so we can have a much faster model evaluation.

**Comment by wanmok** [comment]
@zhuohan123  @WoosukKwon - hi guys, I wonder if we could have someone review the PR. If any change requested, I would be happy to work on it. Thanks!

File: vllm/engine/llm_engine.py [hunk 1]
```diff
@@ -255,6 +255,9 @@ def add_request(
         if prompt_token_ids is None:
             assert prompt is not None
             prompt_token_ids = self.tokenizer.encode(prompt)
+        if prompt is None:
+            prompt = self.tokenizer.decode(prompt_token_ids,
+                                           skip_special_tokens=True)
```
Comment (zhuohan123) [inline vllm/engine/llm_engine.py:260]: Why is this necessary?

File: vllm/entrypoints/openai/api_server.py [hunk 2]
```diff
@@ -426,9 +430,10 @@ async def create_completion(request: CompletionRequest, raw_request: Request):
             top_k=request.top_k,
             stop=request.stop,
             ignore_eos=request.ignore_eos,
-            max_tokens=request.max_tokens,
+            max_tokens=request.max_tokens if not echo_self else 1,
```
Comment (zhuohan123) [inline vllm/entrypoints/openai/api_server.py:433]: I believe `max_tokens` is not in conflict with `echo`, so this `if` should not exist. We should be able to generate many tokens while returning the logits of the prompt.

File: vllm/sequence.py [hunk 3]
```diff
@@ -52,6 +53,9 @@ class SequenceData:
         prompt_token_ids: The token IDs of the prompt.
         output_token_ids: The token IDs of the output.
         cumulative_logprob: The cumulative log probability of the output.
+        prompt_logprobs: The log probabilities of the prompt tokens.
+        prompt_top_logprobs: The log probabilities of the top probability
+        tokens in the prompt.
```
Comment (zhuohan123) [inline vllm/sequence.py:58]: Can we have only one variable `prompt_logprobs`? similar to `output_logprobs` in the `Sequence` class.

File: vllm/sequence.py [hunk 4]
```diff
@@ -62,6 +66,14 @@ def __init__(
         self.output_token_ids: List[int] = []
         self.cumulative_logprob = 0.0
 
+        self.prompt_logprobs: Optional[List[Optional[float]]] = None
+        self.prompt_top_logprobs: Optional[List[Optional[Dict[int,
+                                                              float]]]] = None
```
Comment (zhuohan123) [inline vllm/sequence.py:71]: Let's move these two properties under the `Sequence` class instead of the current `SequenceData` class? The `SequenceData` class only keeps the bare minimal set of data to be the input of the neural network. We need to keep this class small since this class may need to be passed through RPC.

File: vllm/sampling_params.py [hunk 5]
```diff
@@ -51,23 +51,22 @@ class SamplingParams:
...
+                 top_p: float = 1.0,
+                 top_k: int = -1,
+                 use_beam_search: bool = False,
+                 length_penalty: float = 1.0,
+                 early_stopping: Union[bool, str] = False,
+                 stop: Union[None, str, List[str]] = None,
+                 ignore_eos: bool = False,
+                 max_tokens: int = 16,
+                 logprobs: Optional[int] = None,
+                 echo: bool = False) -> None:
```
Comment (zhuohan123) [inline vllm/sampling_params.py:69]: Can we change `echo` to `get_prompt_logprobs` except for the openai API server? This name is easier for a new comer to understand the functionality.
```suggestion
                 get_prompt_logprobs: bool = False) -> None:
```

File: vllm/sequence.py [hunk 6]
```diff
@@ -338,14 +374,18 @@ def __init__(
...
     def __repr__(self) -> str:
         return (f"SequenceOutputs(parent_seq_id={self.parent_seq_id}, "
                 f"output_token={self.output_token}), "
-                f"logprobs={self.logprobs}")
+                f"logprobs={self.logprobs})")
 
     def __eq__(self, other: object) -> bool:
         if not isinstance(other, SequenceOutputs):
-            return NotImplementedError()
+            raise NotImplementedError
```
Comment (zhuohan123) [inline vllm/sequence.py:388]: Thanks for catching this! Let me move this fix to a separate PR.

File: vllm/worker/worker.py [hunk 7]
```diff
@@ -187,6 +188,9 @@ def _prepare_inputs(
                 slot = block_number * self.block_size + block_offset
                 slot_mapping.append(slot)
 
+            # Whether to use echo for prompts.
+            use_echo.append(seq_group_metadata.sampling_params.echo)
```
Comment (zhuohan123) [inline vllm/worker/worker.py:192]: Again, change `use_echo` to `get_prompt_logprobs` can be much more straightforward

File: vllm/model_executor/layers/sampler.py [hunk 8]
```diff
@@ -83,10 +113,93 @@ def forward(
         # Compute the probabilities.
         probs = torch.softmax(logits, dim=-1, dtype=torch.float)
         # Compute the log probabilities (before applying top-p and top-k).
-        logprobs = torch.log(probs)
+        # Re-compute to ensure numerical stability.
+        logprobs = torch.log_softmax(logits, dim=-1, dtype=torch.float)
```
Comment (zhuohan123) [inline vllm/model_executor/layers/sampler.py:117]: Again, thanks for catching this. Let me add this in another PR.

File: vllm/model_executor/layers/sampler.py [hunk 9]
```diff
@@ -83,10 +113,93 @@ def forward(
...
+        prompt_index_list: List[int] = []
+        seq_data = []
+        has_prompt_seq_group_ids = []
+        for i, (x, _) in enumerate(input_metadata.seq_groups):
+            if len(input_metadata.seq_data[x[0]].output_token_ids) == 0:
+                seq_data.append(input_metadata.seq_data[x[0]])
+                has_prompt_seq_group_ids.append(i)
+        prompt_id_list: List[int] = []
+        start_idx = 0
+        for prompt_len, seq_datum in zip(input_metadata.prompt_lens, seq_data):
```
Comment (zhuohan123) [inline vllm/model_executor/layers/sampler.py:147]: Nit: rename `seq_data` to `seq_datas` and rename `seq_datum` to `seq_data`?

File: vllm/model_executor/layers/sampler.py [hunk 10]
```diff
@@ -83,10 +113,93 @@ def forward(
...
         # Compute the log probabilities (before applying top-p and top-k).
-        logprobs = torch.log(probs)
+        # Re-compute to ensure numerical stability.
+        logprobs = torch.log_softmax(logits, dim=-1, dtype=torch.float)
 
         # Sample the next tokens.
-        return _sample(probs, logprobs, input_metadata)
+        samples = _sample(probs, logprobs, input_metadata)
+        if echo_results:
+            for i, ss in enumerate(samples):
```
Comment (zhuohan123) [inline vllm/model_executor/layers/sampler.py:122]: Please avoid confusing variable names like `ss`. Please rename it to `sample` for example.

File: vllm/model_executor/layers/sampler.py [hunk 11]
```diff
@@ -83,10 +113,93 @@ def forward(
...
+                if i in echo_results:
+                    logprobs_list, top_logprobs_list = echo_results[i]
+                    for s in ss:
```
Comment (zhuohan123) [inline vllm/model_executor/layers/sampler.py:125]: Similarly, please avoid using `s` for variable name.

**Review by zhuohan123 — CHANGES_REQUESTED** [review]
Thank you for your contribution! This is a feature that a lot of our community members are asking. The PR is relatively complex and I think it may require several rounds of review. I have left some of my initial comments. Please take a look. Additionally, Is it possible for you to add a test on this functionality [here](https://github.com/vllm-project/vllm/tree/main/tests/samplers)? After this round, I will run this PR, check the correctness of the PR, and then perform another round of review. Thanks again!

File: vllm/entrypoints/openai/api_server.py (hunk 2 above)
Comment (wanmok) [reply vllm/entrypoints/openai/api_server.py:433]: This is a use case myself encounters; OpenAI API supports echoing without generating any tokens while in vLLM we assume that we have to generate at least 1 token. This is a workaround to support the same semantics --- asking the engine to generate 1 token and set the `echo_self` flag to true and removing the generated token at the end before sending back the response.

File: vllm/sequence.py (hunk 3 above)
Comment (wanmok) [reply vllm/sequence.py:58]: In OpenAI implementation, they allow `logprobs=0` while getting back the only logprobs chosen for the prompt. If we only keep one variable, we won't be able to get the logprob for the chosen tokens since they might not be in our top-k.

File: vllm/engine/llm_engine.py (hunk 1 above)
Comment (wanmok) [reply vllm/engine/llm_engine.py:260]: If the user only passes `prompt_token_ids`, the output won't be able to echo the prompt in its string form since it is not provided; it isn't the case vice versa.

Comment (wanmok) [reply vllm/engine/llm_engine.py:260]: If the user only passes prompt_token_ids, the output won't be able to echo the prompt in its string form since it is not provided; it isn't the case vice versa.

File: vllm/sequence.py (hunk 3 above)
Comment (zhuohan123) [reply vllm/sequence.py:58]: We support the OpenAI's behavior and [this is what we did](https://github.com/vllm-project/vllm/blob/d6545ad22ee89c8b6e3eb6cfcf8ff914a06ccee1/vllm/model_executor/layers/sampler.py#L383-L393) for `output_logprobs`: We will select the top-k (can be top-0), and also force the generated token to be in the `output_logprobs` dict. I believe we can do the same thing for `prompt_logprobs`.

File: vllm/sequence.py (hunk 4 above)
Comment (wanmok) [reply vllm/sequence.py:71]: This is actually a workaround that I couldn't figure out a better solution without more intrusive modifications. The way I implemented getting logprobs is to gather these prompt logprobs the first time calling sampler. However, from the view of `Sampler`, the `forward()` method doesn't have access to `Sequence`, making it unable to pass the gathered prompt logprobs back to the `Sequence`. 

One improvement I can think of is to keep these fields in `SequenceData` and make them also available in `Sequence`. Once the `fork()` method finds that they are set in corresponding `SequenceData`, it can pull them out from the `SequenceData` to `Sequence` so that they won't be further passed via network. Please let me know if you got a better thought on that.

File: vllm/entrypoints/openai/api_server.py (hunk 2 above)
Comment (zhuohan123) [reply vllm/entrypoints/openai/api_server.py:433]: Correct me if I'm wrong, but I think what this line does is when `echo` is on, `max_tokens` have to be 1?

File: vllm/sequence.py (hunk 3 above)
Comment (wanmok) [reply vllm/sequence.py:58]: But this might result in different sizes of `logprobs`? Say if we want `logprobs=2`, then for some tokens, it might return 3? or we want another filter at the end?

Comment (zhuohan123) [reply vllm/sequence.py:58]: > if we want logprobs=2, then for some tokens, it might return 3

Yes, and this is OpenAI's behavior as well.

File: vllm/entrypoints/openai/api_server.py (hunk 2 above)
Comment (wanmok) [reply vllm/entrypoints/openai/api_server.py:433]: Nope, it means when `echo==True and max_tokens==0`. Please see line 379.

File: vllm/sequence.py (hunk 3 above)
Comment (wanmok) [reply vllm/sequence.py:58]: I just verified, it isn't. 

Request:
```
{
    "model": "text-davinci-003",
    "prompt": "Please get what I booked",
    "echo": true,
    "n": 1,
    "temperature": 1.0,
    "max_tokens": 1,
    "logprobs": 2
}
```

Response:
```
{
    "object": "text_completion",
    "created": 1694560209,
    "model": "text-davinci-003",
    "choices": [
        {
            "text": "Please get what I booked\n",
            "index": 0,
            "logprobs": {
                "tokens": [
                    "Please",
                    " get",
                    " what",
                    " I",
                    " booked",
                    "\n"
                ],
                "token_logprobs": [
                    null,
                    -6.516949,
                    -7.269826,
                    -1.5403796,
                    -11.766179,
                    -0.064088084
                ],
                "top_logprobs": [
                    null,
                    {
                        " enter": -1.8900735,
                        " note": -2.4450452
                    },
                    {
                        " in": -1.1974856,
                        " your": -2.421246
                    },
                    {
                        " I": -1.5403796,
                        " you": -0.7000267
                    },
                    {
                        " need": -1.2898092,
                        "'m": -1.5458908
                    },
                    {
                        "\n": -0.064088084,
                        " for": -3.8346472
                    }
                ],
                "text_offset": [
                    0,
                    6,
                    10,
                    15,
                    17,
                    24
                ]
            },
            "finish_reason": "length"
        }
    ],
    "usage": {
        "prompt_tokens": 5,
        "completion_tokens": 1,
        "total_tokens": 6
    }
}
```

The prompt tokens I used are not in the returned `top_logprobs` while they are fixed at the size of 2.

Comment (zhuohan123) [reply vllm/sequence.py:58]: I see. We implemented vLLM based on [the description in the API doc](https://platform.openai.com/docs/api-reference/completions/create#logprobs).

However, I believe two variables introduce unnecessary complexity. Let's just use one variable in the code. If we really want to be exactly the same as OpenAI's current behavior, let's maybe add a filter in the end as you suggested?

File: vllm/sequence.py (hunk 4 above)
Comment (zhuohan123) [reply vllm/sequence.py:71]: Instead of directly modifying these fields in the `Sampler`, can we return the probabilities in the `SequenceOutputs` class and modify the `Sequence` class later?

Comment (wanmok) [reply vllm/sequence.py:71]: Ah, I suppose that I have already implemented that.... 🤣 the thing I described was in a previous implementation.... 

Ok, I'm gonna move it to `Sequence`. It should work directly.

File: vllm/entrypoints/openai/api_server.py (hunk 2 above)
Comment (zhuohan123) [reply vllm/entrypoints/openai/api_server.py:433]: got it. Thanks!

**Comment by wanmok** [comment]
@zhuohan123 I have made necessary changes per the review and added a test case to ensure getting the same prompt logprobs as running HF models. Please let me know if you got any other questions.

**Comment by wanmok** [comment]
@zhuohan123 I wonder if there is any update on the review. Please let me know if you found any blocker.

**Review by wanmok — COMMENTED** [review]
I have addressed comments from the previous review. Please let me know if there is any other concerns to merge the PR.

**Comment by lifengjin** [comment]
Any update on this? This is very useful.

**Comment by zhuohan123** [comment]
@wanmok Thanks for all of your contributions so far. I have been looking into this last week and tried to have a simpler modification to the vLLM core to implement this function. Please find my changes in this PR: https://github.com/vllm-project/vllm/pull/1328. Is it possible for you to add the OpenAI endpoint support of `echo` on top of that PR? Let me know if there is any issue.

**Comment by wanmok** [comment]
> @wanmok Thanks for all of your contributions so far. I have been looking into this last week and tried to have a simpler modification to the vLLM core to implement this function. Please find my changes in this PR: #1328. Is it possible for you to add the OpenAI endpoint support of `echo` on top of that PR? Let me know if there is any issue.

Well.... it might take some time to read through the PR and re-implement that. Are you thinking about a new PR after that one being merged or?

**Comment by zhuohan123** [comment]
> Well.... it might take some time to read through the PR and re-implement that. Are you thinking about a new PR after that one being merged or?

PR #1328 modified the vLLM core. So that you can just add `prompt_logprobs=N` in `SamplingParams`, and the returning `RequestOutput` will include a `prompt_logprobs` field in it. After #1328 being merged into the main branch, the only remaining thing is to let the OpenAI API server uses this new feature to support `echo`. Can you submit a new PR for this after #1328 is merged? I believe the code change to `openai/api_server.py` should be very similar with what you have done in this PR.

**Comment by wanmok** [comment]
Sounds good. I can file another PR for OpenAI server later. Also, since I have been working on a similar implementation in this PR, could you list me as co-author for #1328?

-Yunmo

**Comment by zhuohan123** [comment]
Sure thing! Actually in that PR I used some of your testing code. Will add it when I merge the PR.

**Comment by RanchiZhao** [comment]
Cool！is this now available， I mean the logps of prompt tokens.

**Comment by zhuohan123** [comment]
@wanmok #1328 has been merged to the main branch!

**Comment by wanmok** [comment]
> @wanmok #1328 has been merged to the main branch!

Cool! I'll draft a new PR when I got time, probably in a week or so (traveling to CA this week).

**Comment by lifengjin** [comment]
Any updates on this, please?

**Comment by zhuohan123** [comment]
Close this PR in favor of #1328 and #1509. Thank you for your contribution!
